chronicle-queue/src/main/scala/akka/stream/alpakka/chroniclequeue/impl/ChronicleQueueSource.scala

```diff
@@ -0,0 +1,129 @@
+
+  protected def autoCommit(index: Long) = queue.commit(defaultOutputPort, index, false)
+
+  protected def verifyCommit(index: Long) = queue.verifyCommitOrder(defaultOutputPort, index)
+
+  override def createLogic(inheritedAttributes: Attributes): GraphStageLogic =
+    new GraphStageLogic(shape) {
+
+      private val successCallback = getAsyncCallback[Event[T]](handleSuccess)
+
+      private val failureCallback = getAsyncCallback[Exception](handleFailure)
+
+      val pendingOffers = new PriorityQueue[Event[T]]()(new Ordering[Event[T]] {
+        def compare(x: Event[T], y: Event[T]) = // inverse order
+          Ordering.Long.compare(y.index, x.index)
+      })
+
+      var terminating = false
+
+      override def postStop(): Unit =
+        terminating = true
+
+      def receiveMessages(): Unit =
+        Future { queue.dequeue() }(sameThreadExecutionContext).onComplete {
+          case Success(el) =>
+            el match {
+              case Some(element) =>
+                successCallback.invoke(element)
+              case None =>
+                successCallback.invoke(null.asInstanceOf[Event[T]])
+            }
+          case Failure(e) =>
+            e match {
+              case ex: Exception =>
+                failureCallback.invoke(ex)
+              case th: Throwable => throw th
+            }
+        }(sameThreadExecutionContext)
+
+      def handleFailure(ex: Exception): Unit =
+        failStage(ex)
+
+      def commitOrEnqueue(elem: Event[T]): Boolean =
+        try {
+          verifyCommit(elem.index)
+          push(out, elementOut(elem))
+          autoCommit(elem.index)
+          true
+        } catch {
+          case _: Throwable =>
+            pendingOffers.enqueue(elem)
+            false
+        }
+
+      def handleSuccess(result: Event[T]): Unit = {
+        if (!pendingOffers.isEmpty && isAvailable(out)) {
+          if (result != null)
+            pendingOffers.enqueue(result)
+
+          commitOrEnqueue(pendingOffers.dequeue)
+        } else if (pendingOffers.isEmpty && isAvailable(out) && result != null) {
+          commitOrEnqueue(result)
+        } else if (pendingOffers.isEmpty && !isAvailable(out) && result != null) {
+          pendingOffers.enqueue(result)
+        }
+
+        receiveMoreOrComplete()
+      }
+
+      private def receiveMoreOrComplete(): Unit =
+        if (!terminating) {
+          receiveMessages()
+        } else if (terminating) {
+          completeStage()
+        }
+
+      setHandler(
+        out,
+        new OutHandler {
+          override def onPull(): Unit =
+            if (!pendingOffers.isEmpty) {
+              var commited = commitOrEnqueue(pendingOffers.dequeue)
+              while (!pendingOffers.isEmpty && isAvailable(out) && commited) {
+                commited = commitOrEnqueue(pendingOffers.dequeue)
+              }
+              receiveMoreOrComplete()
+            } else {
+              receiveMessages()
```

Comment:
One possible issue I see here, is that when the Queue is empty, this would be a continuous loop of polling the queue with `receiveMessages` -> `dequeue` -> `successCallback.invoke(null.asInstanceOf[Event[T]])` -> `handleSuccess` -> `receiveMoreOrComplete` -> `receiveMessages`. Is this intended?

Reply:
That was one of the reasons we left the source and the sink coupled as a flow. Attempts to decouple have been a little less than successful.

Reply:
Yes, this is intended, and the decoupling from a strict loop is done with the async callback.
Note that everything is also happening on the very same thread, It is my understanding from documentation and examples that this is the intended usage of Chronicle in interprocess communication.
Maybe @lburgazzoli can chip in :pray: .

Reply:
@andreaTP yes, that's correct

Reply:
Do we know the cost of that continuous polling? Looks like the polling thread will stay very busy in a relatively idle queue. Of course, we need to do so in exchange with data latency.

Reply:
Not sure about the cost, we can consider adding a configurable [Pauser](https://github.com/OpenHFT/Chronicle-Queue/blob/master/docs/FAQ.adoc#with-a-tight-reader-loop-i-see-100-utilization-will-there-be-processing-capability-left-for-anything-else) ?